libs/base/Data/String.idr

```diff
@@ -51,9 +51,8 @@ fastUnlines = fastConcat . unlines'
 ||| ```idris example
 ||| words' (unpack " A B C  D E   ")
 ||| ```
-covering
 words' : List Char -> List (List Char)
-words' s = case dropWhile isSpace s of
+words' s = assert_total $ case dropWhile isSpace s of
             [] => []
             s' => let (w, s'') = break isSpace s'
                   in w :: words' s''
```

Comment:
```suggestion
                  in w :: words' (assert_smaller s s'')
```

Reply:
I thought about this too. However, strictly speaking, `s''` is not necessarily smaller than `s`. This is only true after removing the leading space(s) (left over from the call to `break`), which happens, when `dropWhile isSpace` is invoked. So, while it is true that `words'` is total, it is not true that `s''` is always smaller than `s`.

Reply:
So, I dropped the sledgehammer altogether and wrote a tail recursive implementation, which satisfies the totality checker without cheating.

Reply:
In `let (w,s'') = break isSpace s'`, we have that either `s'' = (' ' :: cs)` or `s'' = []`. Could we just drop the first character of `s''` before recurring?

Reply:
Sure, but since this is already done when calling `dropWhile isSpace`, I didn't think it was worth the hassle. Anyway, I completely rewrote the implementation, as I think it's preferable to have a tail call optimized version anyway.